src/org/labkey/test/tests/SampleSetRemoteAPITest.java

```diff
@@ -181,6 +196,328 @@ public void sampleSetWithMissingValueField() throws IOException, CommandExceptio
+                "        dataFileField: 'S3',\n" +
+                "        sampleId: 'def'\n" +
+                "      }]\n" +
+                "    }]\n" +
+                "  }\n" +
+                "})";
+        executeScript(createScript);
+
+        // add a bodyWebPart here to make viewing the sampleset easier while debugging
+        new PortalHelper(getDriver()).addBodyWebPart("Sample Sets");
+        refresh();
+        DataRegionTable sampleSetList =  DataRegionTable.DataRegion(getDriver()).withName("SampleSet").waitFor();
+        waitAndClick(Locator.linkWithText(missingValueTable));
+        DataRegionTable materialsList =  DataRegionTable.DataRegion(getDriver()).withName("Material").waitFor();
+
+        // grab the new samples
+        SelectRowsResponse sampleSetResponse = dgen.getRowsFromServer(createDefaultConnection(true));
+        Map<String, Object> firstAddedSample = sampleSetResponse.getRows()
+                .stream()
+                .filter(a-> a.get("Name").equals("another new one"))
+                .findFirst().orElse(null);;
+        Map<String, Object> secondAddedSample = sampleSetResponse.getRows()
+                .stream()
+                .filter(a-> a.get("Name").equals("a new one"))
+                .findFirst().orElse(null);;
+
+        assertNotNull("expect new samples as outputs from batch save", firstAddedSample);
+        assertNotNull("expect new samples as outputs from batch save", secondAddedSample);
+        int secondAddedRowId = (Integer)secondAddedSample.get("RowId");
+        sleep(2000); // ugh, give it a moment
+
+        // now update one of the new rows and add another
+        String updateScript = "LABKEY.Experiment.saveBatch({\n" +
+                "  assayId: "+assayIdStringValue+",\n" +
+                "  batch: {\n" +
+                "    name: 'foo',\n" +
+                "    runs: [{\n" +
+                "      name: 'bar2',\n" +
+                "\n" +
+                "      materialOutputs: [\n" +
+                "        // create/update samples in the target sample set\n" +
+                "        { rowId: "+secondAddedRowId+", sampleSet: { name: 'mvAssaySamples' }, properties: { 'mvStringData': 'newer from api', volume: 7 } },\n" +
+                "        { name: 'another really new one', sampleSet: { name: 'mvAssaySamples' }, properties: { 'mvStringData': 'also new from api', volume:17 } }\n" +
+                "      ],\n" +
+                "\n" +
+                "      properties: {\n" +
+                "        runFileField: 'blah.png'\n" +
+                "      },\n" +
+                "\n" +
+                "      dataRows: [{\n" +
+                "        dataFileField: 'S4',\n" +
+                "        sampleId: 'def'\n" +
+                "      }]\n" +
+                "    }]\n" +
+                "  }\n" +
+                "})";
+        executeScript(updateScript);
+        sleep(2500); // ugh, again.
```

Comment:
Not sure why you need the sleep, could it be because the update isn't working as expected?
How about calling executeAsyncScript?

Reply:
To use `executeAsyncScript`, you need to provide success and failure callbacks to the `saveBatch` JavaScript config.
Should be as simple as adding this after the `assayId` line in the `updateScript`:
```
"success: callback, failure:callback, \n" +
```
Check the JavaScript docs for more info on [`LABKEY.Experiment.saveBatch(config)`](https://www.labkey.org/download/clientapi_docs/javascript-api/symbols/LABKEY.Experiment.html#.saveBatch)

Reply:
fixed